### DISCUSSION TITLE
von Neuman boundary condition vs free boundary condition

### DISCUSSION THREAD

**Discussion by turczyneq** [kinnala/scikit-fem#1151] [Q&A] [answered]
I am working on example 50 - Advection-diffusion in non-uniform flow. 

I have the following mesh:
![image](https://github.com/user-attachments/assets/d2ca0910-22e8-4cde-b722-56ba31117b2d)

Then I apply boundary conditions as follows(skipped definition of laplacian and avection term for better visibility, they are some BilinearForms):
```
basis = Basis(mesh, ElementTriP1())
...
# Assemble the system matrix
A = asm(claplace, basis) + peclet * asm(advection, basis)

# Identify the interior degrees of freedom
interior = basis.complement_dofs(basis.get_dofs({"bottom", "ball"}))

# Boundary condition
u = basis.zeros()
u[basis.get_dofs("bottom")] = 1.0
u[basis.get_dofs("ball")] = 0.0

u = solve(*condense(A, x=u, I=interior))
```

I wanted to ask about the way to set the boundary condition $\partial_r u(r=0) = 0$ for "left" boundary. Then I found https://github.com/kinnala/scikit-fem/discussions/1084. I noticed that the solution requires setting $\partial_r u = 0$ on boundaries 1, 2, 3, 5, and 6. Despite this requirement, I cannot find any code implementing it. Is it the default behavior?
 - If yes then how to allow boundaries "top" and "right" to have nonzero $\partial_n u$?
 - If no then how to set $\partial_n u = 0$ on "left" boundary?
 
Thank you for the help

**Comment by kinnala** [comment, marked as answer]
Neumann boundary condition is natural (i.e. automatically satisfied) for the variational formulation you have chosen to solve. This is not specific to scikit-fem or even to the finite element method, but rather a general feature of the particular variational formulation which is being solved.

There is lots of literature available on other types of boundary conditions, and I won’t repeat everything here, but you can for example look for ”inhomogeneous Neumann condition” or ”Robin boundary condition”. However, perhaps it helps you to know that there is no specific function in scikit-fem to set other types of boundary conditions: everything is done either via condense() or by adding correct terms to the bilinear or linear forms.

**Reply by turczyneq** [reply]
Thank you
